lib/cassava/client.rb

```diff
@@ -0,0 +1,166 @@
+module Cassava
+  class Client
+    attr_reader :session
+
+    # @param [Cassandra::Session] The session object
+    def initialize(session)
+      @session = session
+    end
+
+    def insert_async(table, data)
+      session.execute_async(insert_statement(table, data), :arguments => data.values)
+    end
+
+    def insert(table, data)
+      session.execute(insert_statement(table, data), :arguments => data.values)
+    end
+
+    # @param [Symbol] table the table name
+    # @param [Array<Symbol>] An optional list of column names (as symbols), to only select those columns
+    # @return [StatementBuilder] A select builder representing the partially completed statement.
+    def select(table, columns = nil)
+      StatementBuilder.new(session).select(table, columns)
+    end
+
+    def delete(table, columns = nil)
+      StatementBuilder.new(session).delete(table, columns)
+    end
+
+    private
+
+    def insert_statement(table, data)
+      column_names = data.keys
+      statement_cql = "INSERT INTO #{table} (#{column_names.join(', ')}) VALUES (#{column_names.map { |x| '?' }.join(',')})"
+      session.prepare(statement_cql)
+    end
+  end
+
+  class StatementBuilder
+    attr_reader :session, :table, :clauses
+
+    CLAUSE_ORDERING = {
+                       :main => 0,
+                       :from => 1,
+                       :where => 2,
+                       :order => 3,
+                       :limit => 4,
+                       :allow_filtering => 5
+                      }
+
+    def initialize(session, clauses = {})
+      @session = session
+      @table = table
+      @clauses = clauses
+    end
+
+    def execute_async(opts = {})
+      session.execute_async(statement, opts)
+    end
+
+    def execute(opts = {})
+      session.execute(statement, opts)
+    end
+
+    def statement
+      clauses.sort_by { |s| CLAUSE_ORDERING[s[0]] }.map { |s| s[1] }.join(' ')
+    end
+
+    def select(table, columns = nil)
+      add_clause(SelectClause.new(table, columns), :main)
+    end
+
+    def delete(table, columns = nil)
+      add_clause(DeleteClause.new(table, columns), :main)
+    end
+
+    def where(arg)
+      clause = clauses[:where] || WhereClause.new([])
+      add_clause(clause.where(arg), :where)
+    end
+
+    def allow_filtering
+      add_clause('ALLOW FILTERING', :allow_filtering)
+    end
+
+    def order(clustering_column, direction = :asc)
+      add_clause("ORDER BY #{clustering_column.to_s} #{direction.to_s}", :order)
+    end
+
+    def limit(n)
+      add_clause("LIMIT #{n}", :limit)
```

Comment:
this is not handling the last_page token correctly. Im pretty sure this should cause a return of a Cassandra::Result::Paged results object